File: 235.cpp
#include <iostream>
#include <vector>

using namespace std;
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (p->val > q->val) swap(p, q);
        if(root->val < p->val){          //p和q都在右子树上
            return lowestCommonAncestor(root->right, p, q);
        }else if(root->val > q->val) {   //p和q都在左子树上
            return lowestCommonAncestor(root->left, p, q);
        }
        return root;        
    }

    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (p->val > q->val) swap(p, q);
        TreeNode* node = root;
        while (node) {
            if(node->val < p->val){          //p和q都在右子树上
                node = node->right;
            }else if(node->val > q->val) {   //p和q都在左子树上
                node = node->left;
            }else {
                return node;
            }    
        }
        return nullptr;        
    }
};

int main(int argc, char const *argv[])
{
    cout << 0;
    return 0;
}

File: 235.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if p.val > q.val:
            p, q = q, p
        while True:
            if p.val <= root.val <= q.val:
                return root
            elif root.val < p.val:
                root = root.right
            else:
                root = root.left


if __name__ == "__main__":
    s = Solution()
    print()
